devtools/replay/RecordingState.jsm

```diff
@@ -0,0 +1,95 @@
+"use strict";
+
+ChromeUtils.defineModuleGetter(
+  this,
+  "Services",
+  "resource://gre/modules/Services.jsm"
+);
+
+const env = Cc["@mozilla.org/process/environment;1"].getService(
+  Ci.nsIEnvironment
+);
+
+const RECORDING_STATE = {
+  ERROR: 0,
+  READY: 1,
+  STARTING: 2,
+  RECORDING: 3,
+  STOPPING: 4,
+};
+
+class RecordingState {
+  constructor (start, stop) {
+    this.recordingMap = new WeakMap();
+    this.startCallback = start;
+    this.stopCallback = stop;
+  }
+
+  // Return whether all tabs are automatically being recorded.
+  isRecordingAllTabs() {
+    return env.get("RECORD_ALL_CONTENT")
+        || Services.prefs.getBoolPref("devtools.recordreplay.alwaysRecord");
+  }
+
+  toggle (browser, button) {
+    const obj = this.recordingMap.get(browser);
+
+    if (!obj || obj.state === RECORDING_STATE.READY) {
+      this._setRecordingState(browser, {
+        ...obj,
+        button,
+        state: RECORDING_STATE.STARTING
+      });
+
+      browser.ownerGlobal.setTimeout(() => this._startRecording(browser), 100);
```

Comment:
This is a bit weird but I couldn't find another reference to `setTimeout` which was available from `DevToolsStartup` but not here. Suggestions welcome!

Reply:
What's the reason for the delay?

Reply:
Without the delay, starting a recording never shows the spinner. Seems that the UI thread is blocked.

Reply:
Tested a bit further and I've removed the timeout from the stop branch but I think it's still good to keep in the start side. I also tried `requestAnimationFrame` but that wasn't a sufficient yield to allow the spinner to show during the processing.